core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -394,53 +396,68 @@ private long checkDeletedAndGCed(VersionValue versionValue) {
     }
 
     @Override
-    public boolean index(Index index) {
-        final boolean created;
+    public void index(Index index) {
         try (ReleasableLock lock = readLock.acquire()) {
             ensureOpen();
             if (index.origin().isRecovery()) {
                 // Don't throttle recovery operations
-                created = innerIndex(index);
+                innerIndex(index);
             } else {
                 try (Releasable r = throttle.acquireThrottle()) {
-                    created = innerIndex(index);
+                    innerIndex(index);
                 }
             }
-        } catch (IllegalStateException | IOException e) {
-            try {
-                maybeFailEngine("index", e);
-            } catch (Exception inner) {
-                e.addSuppressed(inner);
+        } catch (OutOfMemoryError | IllegalStateException t) {
+            index.setFailure(new IndexFailedEngineException(shardId, index.type(), index.id(), t));
+        }
+        handleOperationException("index", index);
+    }
+
+    private void handleOperationException(String source, Operation operation) {
+        if (operation.hasFailure()) {
+            ElasticsearchException failure = operation.getFailure();
+            if (failure.getCause() instanceof IOException ||
+                failure.getCause() instanceof IllegalStateException) {
+                try {
+                    maybeFailEngine(source, (Exception) failure.getCause());
+                } catch (Exception inner) {
+                    operation.getFailure().addSuppressed(inner);
+                }
             }
-            throw new IndexFailedEngineException(shardId, index.type(), index.id(), e);
         }
-        return created;
     }
 
-    private boolean innerIndex(Index index) throws IOException {
+    private void innerIndex(Index index) {
         try (Releasable ignored = acquireLock(index.uid())) {
             lastWriteNanos = index.startTime();
             final long currentVersion;
             final boolean deleted;
             final VersionValue versionValue = versionMap.getUnderLock(index.uid());
             if (versionValue == null) {
-                currentVersion = loadCurrentVersionFromIndex(index.uid());
+                try {
+                    currentVersion = loadCurrentVersionFromIndex(index.uid());
+                } catch (IOException e) {
```

Comment:
Agreed, now the index operation failure is only set on the `index` and `delete` method